cpp/include/cudf/join.hpp

```diff
@@ -367,5 +367,131 @@ std::unique_ptr<cudf::table> cross_join(
+   * from `probe_on` columns. Else, for every column in `probe_on` and `build_on`,
+   * an output column will be produced. For each of these pairs (P, B), P
+   * should exist in `probe_on` and B should exist in `build_on`.
+   * @param compare_nulls Controls whether null join-key values should match or not.
+   * @param mr Device memory resource used to allocate the returned table and columns' device
+   * memory.
+   *
+   * @return Result of joining `build` and `probe` tables on the columns
+   * specified by `build_on` and `probe_on`. The resulting table will be joined columns of
+   * `probe(including common columns)+build(excluding common columns)`.
+   */
+  virtual std::unique_ptr<cudf::table> left_join(
+    cudf::table_view const& probe,
+    std::vector<size_type> const& probe_on,
+    std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+    null_equality compare_nulls         = null_equality::EQUAL,
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource()) const = 0;
+
+  /**
+   * @brief Performs a full join by probing in the internal hash table.
+   *
+   * More details please @see cudf::full_join().
+   *
+   * @param probe The probe table, from which the tuples are probed.
+   * @param probe_on The column indices from `probe` to join on.
+   * @param columns_in_common is a vector of pairs of column indices into
+   * `build` and `probe`, respectively, that are "in common". For "common"
+   * columns, only a single output column will be produced, which is gathered
+   * from `probe_on` columns. Else, for every column in `probe_on` and `build_on`,
+   * an output column will be produced. For each of these pairs (P, B), P
+   * should exist in `probe_on` and B should exist in `build_on`.
+   * @param compare_nulls Controls whether null join-key values should match or not.
+   * @param mr Device memory resource used to allocate the returned table and columns' device
+   * memory.
+   *
+   * @return Result of joining `build` and `probe` tables on the columns
+   * specified by `build_on` and `probe_on`. The resulting table will be joined columns of
+   * `probe(including common columns)+build(excluding common columns)`.
+   */
+  virtual std::unique_ptr<cudf::table> full_join(
+    cudf::table_view const& probe,
+    std::vector<size_type> const& probe_on,
+    std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+    null_equality compare_nulls         = null_equality::EQUAL,
+    rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource()) const = 0;
+
+  /**
+   * @brief Instantiates a hash join instance and build the internal hash table used for subsequent
+   * probe calls.
+   *
+   * @param build The build table, from which the hash table is built.
+   * @param build_on The column indices from `build` to join on.
+   * @return Hash join instance.
+   */
+  static std::unique_ptr<const hash_join> create(cudf::table_view const& build,
```

Comment:
Why is this a static factory method? I think this should just be the constructor for the `hash_join` class.

Reply:
Also, could we make the constructor take an optional stream argument? Since this PR is going to change the join interface anyway, this seems to be the right chance to expose asynchrony for hash join. @jrhemstad What do you think?

Reply:
The OO `hash_join` interface is still a public interface and we don't want to expose streams in public interfaces yet.

Reply:
The static factory method and the subsequent virtual `*_join` methods are consequences of that I was trying to hide the actual `multimap_type` type and member from the public header file.

But if that is not a concern, I'll save the aforementioned static/virtual methods. Any suggestions?

Reply:
Ah... I see. That makes more sense now. I agree that it seems worthwhile to avoid exposing the `multimap` type. My preference would be to use PImpl to accomplish the same goal. That way we can avoid the awkwardness of the static factory.

Reply:
Yeah PImpl sounds better. Thanks for reminding, will do.

Reply:
Addressed.